File: Compiler/Generators/TypeScript/BebopView.ts

const hexDigits = "0123456789abcdef";
const asciiToHex = [
    0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
    0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
    0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
    0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  0,  0,  0,  0,  0,  0,
    0, 10, 11, 12, 13, 14, 15,  0,  0,  0,  0,  0,  0,  0,  0,  0,
    0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
    0, 10, 11, 12, 13, 14, 15,  0,  0,  0,  0,  0,  0,  0,  0,  0,
    0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0
];

export class PierogiView {
    private buffer: Uint8Array;
    private view: DataView;
    index: number;
    private byteToHex: string[] = []; // A lookup table: ['00', '01', ..., 'ff']
    length: number;

    constructor(data?: Uint8Array) {
        if (data && !(data instanceof Uint8Array)) {
            throw new Error("data is not a Uint8Array");
        }
        this.buffer = data || new Uint8Array(256);
        this.view = new DataView(this.buffer.buffer);
        this.index = 0;
        this.length = data ? data.length : 0;
        for (const x of hexDigits) {
            for (const y of hexDigits) {
                this.byteToHex.push(x + y);
            }
        }
    }

    private growBy(amount: number): void {
        if (this.length + amount > this.buffer.length) {
            const data = new Uint8Array(this.length + amount << 1);
            data.set(this.buffer);
            this.buffer = data;
            this.view = new DataView(data.buffer);
        }
        this.length += amount;
    }

    rewind(): void {
        this.index = 0;
        this.length = 0;
    }

    skip(amount: number) {
        this.index += amount;
    }

    toArray(): Uint8Array {
        return this.buffer.subarray(0, this.length);
    }

    readByte():    number { return this.buffer[this.index++]; }
    readUint16():  number { const result = this.view.getUint16(this.index, true);    this.index += 2; return result; }
    readInt16():   number { const result = this.view.getInt16(this.index, true);     this.index += 2; return result; }
    readUint32():  number { const result = this.view.getUint32(this.index, true);    this.index += 4; return result; }
    readInt32():   number { const result = this.view.getInt32(this.index, true);     this.index += 4; return result; }
    readUint64():  bigint { const result = this.view.getBigUint64(this.index, true); this.index += 8; return result; }
    readInt64():   bigint { const result = this.view.getBigInt64(this.index, true);  this.index += 8; return result; }
    readFloat32(): number { const result = this.view.getFloat32(this.index, true);   this.index += 4; return result; }
    readFloat64(): number { const result = this.view.getFloat64(this.index, true);   this.index += 8; return result; }

    writeByte(value: number):    void { const index = this.length; this.growBy(1); this.buffer[index] = value; }
    writeUint16(value: number):  void { const index = this.length; this.growBy(2); this.view.setUint16(index, value, true); }
    writeInt16(value: number):   void { const index = this.length; this.growBy(2); this.view.setInt16(index, value, true); }
    writeUint32(value: number):  void { const index = this.length; this.growBy(4); this.view.setUint32(index, value, true); }
    writeInt32(value: number):   void { const index = this.length; this.growBy(4); this.view.setInt32(index, value, true); }
    writeUint64(value: bigint):  void { const index = this.length; this.growBy(8); this.view.setBigUint64(index, value, true); }
    writeInt64(value: bigint):   void { const index = this.length; this.growBy(8); this.view.setBigInt64(index, value, true); }
    writeFloat32(value: number): void { const index = this.length; this.growBy(4); this.view.setFloat32(index, value, true); }
    writeFloat64(value: number): void { const index = this.length; this.growBy(8); this.view.setFloat64(index, value, true); }

    readBytes(): Uint8Array {
        const length = this.readUint32(), start = this.index, end = start + length;
        this.index = end;
        return this.buffer.subarray(start, end);
    }

    writeBytes(value: Uint8Array): void {
        this.writeUint32(value.length);
        const index = this.length;
        this.growBy(value.length);
        this.buffer.set(value, index);
    }

    /**
     * Reads a null-terminated UTF-8-encoded string.
     */
    readString(): string {
        let result = "";

        while (true) {
            let codePoint: number;

            // decode UTF-8
            const a = this.readByte();
            if (a < 0xC0) {
                codePoint = a;
            } else {
                const b = this.readByte();
                if (a < 0xE0) {
                    codePoint = ((a & 0x1F) << 6) | (b & 0x3F);
                } else {
                    const c = this.readByte();
                    if (a < 0xF0) {
                        codePoint = ((a & 0x0F) << 12) | ((b & 0x3F) << 6) | (c & 0x3F);
                    } else {
                        const d = this.readByte();
                        codePoint = ((a & 0x07) << 18) | ((b & 0x3F) << 12) | ((c & 0x3F) << 6) | (d & 0x3F);
                    }
                }
            }

            // strings are null-terminated
            if (codePoint === 0) {
                break;
            }

            // encode UTF-16
            if (codePoint < 0x10000) {
                result += String.fromCharCode(codePoint);
            } else {
                codePoint -= 0x10000;
                result += String.fromCharCode((codePoint >> 10) + 0xD800, (codePoint & ((1 << 10) - 1)) + 0xDC00);
            }
        }

        return result;
    }

    /**
     * Writes a null-terminated UTF-8-encoded string.
     */
    writeString(value: string): void {
        let codePoint: number;

        for (let i = 0; i < value.length; i++) {
            // decode UTF-16
            const a = value.charCodeAt(i);
            if (i + 1 === value.length || a < 0xD800 || a >= 0xDC00) {
                codePoint = a;
            } else {
                const b = value.charCodeAt(++i);
                codePoint = (a << 10) + b + (0x10000 - (0xD800 << 10) - 0xDC00);
            }

            // strings are null-terminated, so a string cannot contain a null character.
            if (codePoint === 0) {
                throw new Error("Cannot encode a string containing the null character");
            }

            // encode UTF-8
            if (codePoint < 0x80) {
                this.writeByte(codePoint);
            } else {
                if (codePoint < 0x800) {
                    this.writeByte(((codePoint >> 6) & 0x1F) | 0xC0);
                } else {
                    if (codePoint < 0x10000) {
                        this.writeByte(((codePoint >> 12) & 0x0F) | 0xE0);
                    } else {
                        this.writeByte(((codePoint >> 18) & 0x07) | 0xF0);
                        this.writeByte(((codePoint >> 12) & 0x3F) | 0x80);
                    }
                    this.writeByte(((codePoint >> 6) & 0x3F) | 0x80);
                }
                this.writeByte((codePoint & 0x3F) | 0x80);
            }
        }

        // write the null terminator
        this.writeByte(0);
    }

    readGuid(): string {
        // Order: 3 2 1 0 - 5 4 - 7 6 - 8 9 - a b c d e f
        const b = this.byteToHex, a = this.buffer, i = this.index, d = '-';
        var s = b[a[i + 3]];
        s += b[a[i + 2]];
        s += b[a[i + 1]];
        s += b[a[i]];
        s += d;
        s += b[a[i + 5]];
        s += b[a[i + 4]];
        s += d;
        s += b[a[i + 7]];
        s += b[a[i + 6]];
        s += d;
        s += b[a[i + 8]];
        s += b[a[i + 9]];
        s += d;
        s += b[a[i + 10]];
        s += b[a[i + 11]];
        s += b[a[i + 12]];
        s += b[a[i + 13]];
        s += b[a[i + 14]];
        s += b[a[i + 15]];
        this.index += 16;
        return s;
    }

    writeGuid(value: string): void {
        const v = this.view, i = this.length;
        this.growBy(16);
        var p = 0, a = 0;
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        p += (value.charCodeAt(p) === 45) as any;
        v.setUint32(i, a, true);
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        p += (value.charCodeAt(p) === 45) as any;
        v.setUint16(i + 4, a, true);
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        p += (value.charCodeAt(p) === 45) as any;
        v.setUint16(i + 6, a, true);
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        p += (value.charCodeAt(p) === 45) as any;
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        v.setUint32(i + 8, a, false);
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        a = a<<4 | asciiToHex[value.charCodeAt(p++)];
        v.setUint32(i + 12, a, false);
    }

    writeEnum(value: any): void {
        var encoded = value as number;
        if (encoded === void 0) throw new Error("Couldn't convert enum value");
        this.writeUint32(encoded);
    }

    /**
     * Reserve some space to write a message's length prefix, and return its index.
     * The length is stored as a little-endian fixed-width unsigned 32-bit integer, so 4 bytes are reserved.
     */
    reserveMessageLength(): number {
        const i = this.length;
        this.growBy(4);
        return i;
    }

    /**
     * Fill in a message's length prefix.
     */
    fillMessageLength(position: number, messageLength: number): void {
        this.view.setUint32(position, messageLength, true);
    }

    /**
     * Read out a message's length prefix.
     */
    readMessageLength(): number {
        const result = this.view.getUint32(this.index, true);
        this.index += 4;
        return result;
    }
}

File: Laboratory/C#/BebopView.cs
using System;
using System.Buffers;
using System.Runtime.CompilerServices;
using System.Text;
using Bebop.Exceptions;
using static System.Runtime.CompilerServices.Unsafe;
using static System.Runtime.InteropServices.MemoryMarshal;

namespace Bebop
{
    /// <summary>
    ///     A low-level interface for encoding and decoding Bebop structured data
    ///     TODO disable big-endian systems
    /// </summary>
    public ref struct BebopView
    {
        private static readonly UTF8Encoding UTF8 = new UTF8Encoding();


        const int MaxStackSize = 256;
#if AGGRESSIVE_OPTIMIZE
        public const MethodImplOptions HotPath =
            MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization;
    #else
        public const MethodImplOptions HotPath = MethodImplOptions.AggressiveInlining;
    #endif

        /// <summary>
        ///     A contiguous region of memory that contains the contents of a Bebop message
        /// </summary>
        private Span<byte> _buffer;

        /// <summary>
        ///     Allocates a new <see cref="BebopView"/> instance backed by an empty array.
        /// </summary>
        /// <returns></returns>
        [MethodImpl(HotPath)]
        public static BebopView Create() => new(Array.Empty<byte>());

        [MethodImpl(HotPath)]
        public static BebopView From(Memory<byte> buffer) => From(buffer.Span);

        [MethodImpl(HotPath)]
        public static BebopView From(byte[] buffer) => new(buffer);

        [MethodImpl(HotPath)]
        public static BebopView From(Span<byte> buffer) => new(buffer);


        /// <summary>
        ///     Creates a read-only slice of the underlying <see cref="_buffer"/> containing all currently written data.
        /// </summary>
        [MethodImpl(HotPath)]
        public ReadOnlySpan<byte> Slice() => _buffer.Slice(0, Length);

        /// <summary>
        ///     Copies the contents of <see cref="Slice"/> into a new array.  This heap
        ///     allocates, so should generally be avoided for writing and only used when setting a decoded message property.
        /// </summary>
        [MethodImpl(HotPath)]
        public byte[] ToArray() => Slice().ToArray();

        private BebopView(Span<byte> buffer)
        {
            if (!BitConverter.IsLittleEndian)
            {
                throw new BebopViewException("Big-endian systems are not supported by Bebop.");
            }
            _buffer = buffer;
            Position = 0;
            Length = _buffer.Length;
        }

        /// <summary>
        ///     The current byte position of the view
        /// </summary>
        public int Position { get; set; }

        /// <summary>
        ///     The amount of bytes that have been written to the underlying buffer.
        ///     <remarks>
        ///         This is not the same as the <see cref="_buffer"/> length which contains null-bytes due to look-ahead
        ///         allocation.
        ///     </remarks>
        /// </summary>
        public int Length { get; set; }

        [MethodImpl(HotPath)]
        public ushort ReadUInt16()
        {
            const int size = 2;
            var value = ReadUnaligned<ushort>(ref GetReference(_buffer.Slice(Position, size)));
            Position += size;
            return value;
        }

        [MethodImpl(HotPath)]
        public short ReadInt16()
        {
            const int size = 2;
            var value = ReadUnaligned<short>(ref GetReference(_buffer.Slice(Position, size)));
            Position += size;
            return value;
        }

        [MethodImpl(HotPath)]
        public uint ReadUInt32()
        {
            const int size = 4;
            var value = ReadUnaligned<uint>(ref GetReference(_buffer.Slice(Position, size)));
            Position += size;
            return value;
        }

        [MethodImpl(HotPath)]
        public int ReadInt32()
        {
            const int size = 4;
            var value = ReadUnaligned<int>(ref GetReference(_buffer.Slice(Position, size)));
            Position += size;
            return value;
        }

        [MethodImpl(HotPath)]
        public ulong ReadUInt64()
        {
            const int size = 8;
            var value = ReadUnaligned<ulong>(ref GetReference(_buffer.Slice(Position, size)));
            Position += size;
            return value;
        }

        [MethodImpl(HotPath)]
        public long ReadInt64()
        {
            const int size = 8;
            var value = ReadUnaligned<long>(ref GetReference(_buffer.Slice(Position, size)));
            Position += size;
            return value;
        }

        [MethodImpl(HotPath)]
        public float ReadFloat32()
        {
            const int size = 4;
            var value = ReadUnaligned<float>(ref GetReference(_buffer.Slice(Position, size)));
            Position += size;
            return value;
        }

        [MethodImpl(HotPath)]
        public double ReadFloat64()
        {
            const int size = 8;
            var value = ReadUnaligned<double>(ref GetReference(_buffer.Slice(Position, size)));
            Position += size;
            return value;
        }

        [MethodImpl(HotPath)]
        public void WriteUInt16(ushort value)
        {
            const int size = 2;
            var index = Length;
            GrowBy(size);
            WriteUnaligned(ref GetReference(_buffer.Slice(index, size)), value);
        }

        [MethodImpl(HotPath)]
        private static uint ConvertEnum<T>(T enumValue) where T : struct, Enum =>
           As<T, uint>(ref enumValue);

        [MethodImpl(HotPath)]
        private static T ConvertFrom<T>(uint constant) where T : struct, Enum =>
            As<uint, T>(ref constant);

        [MethodImpl(HotPath)]
        public T ReadEnum<T>() where T : struct, Enum
        {
           return ConvertFrom<T>(ReadUInt32());
        }

        [MethodImpl(HotPath)]
        public void WriteEnum<T>(T value) where T : struct, Enum
        {
            WriteUInt32(ConvertEnum<T>(value));
        }

        [MethodImpl(HotPath)]
        public void WriteFloat32s(float[] value)
        {
            WriteUInt32(unchecked((uint) value.Length));
            var index = Length;
            GrowBy(value.Length);
            AsBytes(value.AsSpan()).CopyTo(_buffer.Slice(index, value.Length));
        }

        [MethodImpl(HotPath)]
        public void WriteFloat64s(double[] value)
        {
            WriteUInt32(unchecked((uint) value.Length));
            var index = Length;
            GrowBy(value.Length);
            AsBytes(value.AsSpan()).CopyTo(_buffer.Slice(index, value.Length));
        }

        [MethodImpl(HotPath)]
        public void WriteInt16(short value)
        {
            const int size = 2;
            var index = Length;
            GrowBy(size);
            WriteUnaligned(ref GetReference(_buffer.Slice(index, size)), value);
        }

        [MethodImpl(HotPath)]
        public void WriteUInt32(uint value)
        {
            const int size = 4;
            var index = Length;
            GrowBy(size);
            WriteUnaligned(ref GetReference(_buffer.Slice(index, size)), value);
        }

        [MethodImpl(HotPath)]
        public void WriteInt32(int value)
        {
            const int size = 4;
            var index = Length;
            GrowBy(size);
            WriteUnaligned(ref GetReference(_buffer.Slice(index, size)), value);
        }

        [MethodImpl(HotPath)]
        public void WriteUInt64(ulong value)
        {
            const int size = 8;
            var index = Length;
            GrowBy(size);
            WriteUnaligned(ref GetReference(_buffer.Slice(index, size)), value);
        }

        [MethodImpl(HotPath)]
        public void WriteInt64(long value)
        {
            const int size = 8;
            var index = Length;
            GrowBy(size);
            WriteUnaligned(ref GetReference(_buffer.Slice(index, size)), value);
        }

        [MethodImpl(HotPath)]
        public void WriteFloat32(float value)
        {
            const int size = 4;
            var index = Length;
            GrowBy(size);
            WriteUnaligned(ref GetReference(_buffer.Slice(index, size)), value);
        }

        [MethodImpl(HotPath)]
        public void WriteFloat64(double value)
        {
            const int size = 8;
            var index = Length;
            GrowBy(size);
            WriteUnaligned(ref GetReference(_buffer.Slice(index, size)), value);
        }

        [MethodImpl(HotPath)]
        public DateTime ReadDate()
        {
            return DateTime.FromBinary(ReadInt64());
        }

        /// <summary>
        ///     Reads a null-terminated string from the underlying buffer
        /// </summary>
        /// <returns>A UTF-16 encoded string</returns>
        [MethodImpl(HotPath)]
        public string ReadString()
        {
            var stringByteCount = unchecked((int) ReadUInt32());
            var stringSpan = _buffer.Slice(Position, stringByteCount);
            Position += stringByteCount;


        #if AGGRESSIVE_OPTIMIZE
            return UTF8.GetString(stringSpan);
        #else
            unsafe
            {
                fixed (byte* bytePtr = stringSpan)
                {
                    return UTF8.GetString(bytePtr, stringByteCount);
                }
            }
        #endif
        }


        [MethodImpl(HotPath)]
        public byte ReadByte() => _buffer[Position++];

        /// <summary>
        ///     Reads a <see cref="Guid"/> from the underlying buffer and advances the advances the current position by 16 bytes.
        /// </summary>
        /// <returns>A well-formed Guid structure instances.</returns>
        [MethodImpl(HotPath)]
        public Guid ReadGuid()
        {
            const int size = 16;
            var index = Position;
            Position += size;
            return ReadUnaligned<Guid>(ref GetReference(_buffer.Slice(index, size)));
        }

        [MethodImpl(HotPath)]
        public void GrowBy(int amount)
        {
            if ((Length & 0xC0000000) != 0)
            {
                throw new BebopViewException("A Bebop View cannot grow beyond 2 gigabytes.");
            }
            if (Length + amount > _buffer.Length)
            {

                var newBuffer = new Span<byte>(new byte[(Length + amount) << 1]);
                _buffer.CopyTo(newBuffer);
                _buffer = newBuffer;
            }
            Length += amount;
        }

        /// <summary>
        ///     Reads a length-prefixed byte array from the buffer
        /// </summary>
        /// <returns>An array of bytes</returns>
        [MethodImpl(HotPath)]
        public byte[] ReadBytes()
        {
            var length = unchecked((int) ReadUInt32());
            var index = Position;
            Position += length;
            return _buffer.Slice(index, length).ToArray();
        }

        [MethodImpl(HotPath)]
        public void WriteBytes(byte[] value)
        {
            WriteUInt32(unchecked((uint) value.Length));
            var index = Length;
            GrowBy(value.Length);
            value.AsSpan().CopyTo(_buffer.Slice(index, value.Length));
        }

        [MethodImpl(HotPath)]
        public void WriteDate(DateTime date)
        {
            WriteInt64(date.ToUniversalTime().ToBinary());
        }

        [MethodImpl(HotPath)]
        public void WriteGuid(Guid guid)
        {
            const int size = 16;
            var index = Length;
            GrowBy(size);
            WriteUnaligned(ref GetReference(_buffer.Slice(index, size)), guid);
        }

        [MethodImpl(HotPath)]
        public void WriteString(string value)
        {
            unsafe
            {
                fixed (char* c = value)
                {
                    var size = UTF8.GetByteCount(c, value.Length);
                    WriteUInt32(unchecked((uint) size));
                    var index = Length;
                    GrowBy(size);
                    fixed (byte* o = _buffer.Slice(index, size))
                    {
                        UTF8.GetBytes(c, value.Length, o, size);
                    }
                }
            }
        }

        [MethodImpl(HotPath)]
        public void WriteByte(bool value)
        {
            const int size = 1;
            var index = Length;
            GrowBy(size);
            _buffer[index] = !value ? 0 : 1;
        }

        [MethodImpl(HotPath)]
        public void WriteByte(byte value)
        {
            const int size = 1;

            var index = Length;
            GrowBy(size);
            _buffer[index] = value;
        }


        /// <summary>
        ///     Reserve some space to write a message's length prefix, and return its index.
        ///     The length is stored as a little-endian fixed-width unsigned 32 - bit integer, so 4 bytes are reserved.
        /// </summary>
        /// <returns>The index to later write the message's length to.</returns>
        [MethodImpl(HotPath)]
        public int ReserveMessageLength()
        {
            const int size = 4;
            var i = Length;
            GrowBy(size);
            return i;
        }

        /// <summary>
        ///     Fill in a message's length prefix.
        /// </summary>
        /// <param name="position">The position in the buffer of the message's length prefix.</param>
        /// <param name="messageLength">The message length to write.</param>
        [MethodImpl(HotPath)]
        public void FillMessageLength(int position, uint messageLength)
        {
            const int size = 4;
            WriteUnaligned(ref GetReference(_buffer.Slice(position, size)), messageLength);
        }

        /// <summary>
        ///     Read out a message's length prefix.
        /// </summary>
        [MethodImpl(HotPath)]
        public uint ReadMessageLength()
        {
            const int size = 4;
            var result = ReadUnaligned<uint>(ref GetReference(_buffer.Slice(Position, size)));
            Position += size;
            return result;
        }
    }
}